src/common/events/events_io.cpp

```diff
@@ -137,6 +137,161 @@ mio::ExpectedSize ReaderFromAsyncReader::Read(
 	return read;
 }
 
+TeeReader::ExpectedTeeReaderLeafPtr TeeReader::AttachSecondaryReader() {
+	if (any_of(
+			leaf_readers_.begin(),
+			leaf_readers_.end(),
+			[](const std::pair<TeeReaderLeafPtr, TeeReaderLeafContext> r) {
+				return r.second.buffer_bytes_missing != 0;
+			})) {
+		return expected::unexpected(error::Error(
+			make_error_condition(errc::io_error), "A Reader is already reading from the buffer"));
+	}
+
+	if (stop_done_) {
+		return expected::unexpected(error::Error(
+			make_error_condition(errc::io_error), "Buffering stopped, no more readers allowed"));
+	}
+
+	auto reader = make_shared<TeeReaderLeaf>(shared_from_this());
+	leaf_readers_.insert({reader, TeeReaderLeafContext {}});
+	auto ex_bytes_missing = source_reader_->Rewind();
+	if (!ex_bytes_missing) {
+		return expected::unexpected(ex_bytes_missing.error());
+	}
+	leaf_readers_[reader].buffer_bytes_missing = ex_bytes_missing.value();
+	return reader;
+}
+
+error::Error TeeReader::ReadyToAsyncRead(
+	TeeReader::TeeReaderLeafPtr leaf_reader,
+	vector<uint8_t>::iterator start,
+	vector<uint8_t>::iterator end,
+	mio::AsyncIoHandler handler) {
+	// The reader must exist in the internal map.
+	auto found = leaf_readers_.find(leaf_reader);
+	AssertOrReturnError(found != leaf_readers_.end());
+
+	if (leaf_readers_[leaf_reader].buffer_bytes_missing > 0) {
+		// Special case, reading missing bytes
+		TeeReaderLeafContext &ctx = leaf_readers_[leaf_reader];
+		auto to_read = std::min(ctx.buffer_bytes_missing, (size_t) (end - start));
+		auto handler_wrapper = [this, handler, &ctx](mio::ExpectedSize result) {
+			if (result) {
+				ctx.buffer_bytes_missing -= result.value();
+			}
+			handler(result);
+			MaybeDiscardBuffer();
+		};
+
+		auto err = source_reader_->AsyncRead(start, start + to_read, handler_wrapper);
+		if (err != error::NoError) {
+			handler(expected::unexpected(err));
+		}
+	} else {
+		leaf_readers_[leaf_reader].pending_read.start = start;
+		leaf_readers_[leaf_reader].pending_read.end = end;
+		leaf_readers_[leaf_reader].pending_read.handler = handler;
+		if (++ready_to_read == leaf_readers_.size()) {
+			DoAsyncRead();
+			ready_to_read = 0;
+		}
+	}
+
+	return error::NoError;
+}
+
+void TeeReader::CallAllHandlers(mio::ExpectedSize result) {
+	// Makes a copy of the handlers and then calls them sequentially
+	vector<mio::AsyncIoHandler> handlers;
+	for (auto const &it : leaf_readers_) {
+		handlers.push_back(it.second.pending_read.handler);
+	}
+	for (auto const &h : handlers) {
+		h(result);
+	}
+}
+
+void TeeReader::DoAsyncRead() {
+	auto handler = [this](mio::ExpectedSize result) {
+		if (!result) {
+			CallAllHandlers(result);
+			return;
+		};
+
+		auto start_iterator = leaf_readers_.begin()->second.pending_read.start;
+		auto read_bytes = result.value();
+		for_each(
+			++leaf_readers_.begin(),
```

Comment:
This is to start in the second position, not the first, right? Could you use `begin() + 1` instead? I don't think `++` has any ill effects here, but it's confusing to read because it also modifies its argument, which is a temporary (I'm mildly surprised it's even allowed here).

Reply:
Right, because you can't jump an arbitrary number of steps, just one at a time. But `std::next` is even more descriptive, so 👍.